runtime/catalog-karapace/src/main/java/io/aklivity/zilla/runtime/catalog/karapace/internal/KarapaceCatalogHandler.java

```diff
@@ -88,21 +100,27 @@ public String resolve(
         }
         else
         {
-            CompletableFuture<String> future = cache.get(schemaId);
+            CompletableFuture<String> future = cachedSchemas.get(schemaId);
             if (future == null)
             {
-                future = CompletableFuture.supplyAsync(() ->
+                CompletableFuture<String> newFuture = new CompletableFuture<String>();
+                future = cachedSchemas.putIfAbsent(schemaId, newFuture);
+                if (future == null)
                 {
-                    String response = sendHttpRequest(MessageFormat.format(SCHEMA_PATH, schemaId));
-                    return response != null ? request.resolveSchemaResponse(response) : null;
-                });
+                    newFuture = CompletableFuture.supplyAsync(() ->
+                    {
+                        String response = sendHttpRequest(MessageFormat.format(SCHEMA_PATH, schemaId));
+                        return response != null ? request.resolveSchemaResponse(response) : null;
+                    });
+                    future = newFuture;
```

Comment:
I don't think we can reassign `newFuture` here, as the value being lost is the one that already made it into the concurrent map, so if we reassign it then the one in the map would not be completed, agree?